code/datums/storage/storage.dm

```diff
@@ -0,0 +1,417 @@
+			if (user.s_active)
+				user.detach_hud(user.s_active)
+				user.s_active = null
+			if (src.mousetrap_check(user))
+				return
+			user.s_active = src.hud
+			src.hud.update(user)
+			user.attach_hud(src.hud)
+			src.linked_item.add_fingerprint(user)
+			animate_storage_rustle(src.linked_item)
+			return
+		else
+			// make sure only the user can see the storage
+			for (var/mob/M as anything in src.hud.mobs)
+				if (M != user)
+					M.detach_hud(hud)
+			src.hud.update(user)
+
+	// storage item is mouse dropped onto something
+	proc/storage_item_mouse_drop(mob/user, atom/over_object, src_location, over_location)
+		// if mouse dropping storage item onto a hand slot, attempt to hold it
+		if (istype(over_object, /atom/movable/screen/hud))
+			var/atom/movable/screen/hud/S = over_object
+			playsound(src.linked_item.loc, "rustle", 50, TRUE, -5)
+			if (!user.restrained() && !user.stat && src.linked_item.loc == user)
+				if (S.id == "rhand" && !user.r_hand)
+					user.u_equip(src.linked_item)
+					user.put_in_hand_or_drop(src.linked_item)
+				else if (S.id == "lhand" && !user.l_hand)
+					user.u_equip(src.linked_item)
+					user.put_in_hand_or_drop(src.linked_item)
+		// if mouse dropping storage item onto self, look inside
+		else if (over_object == user && in_interact_range(src.linked_item, user) && isliving(user) && !user.stat && !isintangible(user))
+			if (user.s_active)
+				user.detach_hud(user.s_active)
+				user.s_active = null
+			if (src.mousetrap_check(user))
+				return
+			user.s_active = src.hud
+			hud.update(user)
+			user.attach_hud(src.hud)
+		// if mouse dropping onto a table or rack, attempt to dump contents onto it
+		else if (user.is_in_hands(src.linked_item))
+			var/turf/T = over_object
+			if (istype(T, /obj/table) || istype(T, /obj/rack))
+				T = get_turf(T)
+			if (!(user in range(1, T)))
+				return
+			if (!istype(T))
+				return
+			if (T.density)
+				return
+			for (var/obj/O in T)
+				if (O.density && !istype(O, /obj/table) && !istype(O, /obj/rack))
+					return
+			user.visible_message("<span class='alert'>[user] dumps the contents of [src.linked_item.name] onto [over_object]!</span>")
+			for (var/obj/item/I as anything in src.get_contents())
+				src.transfer_stored_item(I, T, user = user)
+				I.layer = initial(I.layer)
+				if (istype(I, /obj/item/mousetrap))
+					var/obj/item/mousetrap/MT = I
+					if (MT.armed)
+						MT.visible_message("<span class='alert'>[MT] triggers as it falls on the ground!</span>")
+						MT.triggered(user, null)
+				else if (istype(I, /obj/item/mine))
+					var/obj/item/mine/M = I
+					if (M.armed && M.used_up != TRUE)
+						M.visible_message("<span class='alert'>[M] triggers as it falls on the ground!</span>")
+						M.triggered(user)
+
+	// using storage item in hand
+	proc/storage_item_attack_self(mob/user)
+		src.storage_item_attack_hand(user)
+
+	// after attacking an object with the storage item
+	proc/storage_item_after_attack(atom/target, mob/user, reach)
```

Comment:
please absolute path these procs, for something such core functionality like this we definitely need control-f searchability